file.go

```diff
@@ -7,56 +7,109 @@ import (
-
-	// FullPath returns the full path used when adding with this file
-	FullPath() string
-
-	// IsDirectory returns true if the File is a directory (and therefore
-	// supports calling `NextFile`) and false if the File is a normal file
-	// (and therefor supports calling `Read` and `Close`)
-	IsDirectory() bool
-
-	// NextFile returns the next child file available (if the File is a
-	// directory). It will return (nil, io.EOF) if no more files are
-	// available. If the file is a regular file (not a directory), NextFile
-	// will return a non-nil error.
-	NextFile() (File, error)
+	io.Closer
+
+	// Size returns size of this file (if this file is a directory, total size of
+	// all files stored in the tree should be returned). Some implementations may
+	// choose not to implement this
+	Size() (int64, error)
 }
 
-type StatFile interface {
+// Regular represents a regular Unix file
+type Regular interface {
 	File
 
-	Stat() os.FileInfo
+	io.Reader
+	io.Seeker
+}
+
+// DirEntry exposes information about a directory entry
+type DirEntry interface {
+	// Name returns the base name of this entry, which is the base name of
+	// the referenced file
+	Name() string
+
+	// File returns the file referenced by this DirEntry
+	File() File
+
+	// Regular is an alias for ent.File().(Regular). If the file isn't a regular
+	// file, nil value will be returned
+	Regular() Regular
+
+	// Dir is an alias for ent.File().(directory). If the file isn't a directory,
+	// nil value will be returned
+	Dir() Directory
 }
 
-type PeekFile interface {
-	SizeFile
+// DirIterator is a iterator over directory entries.
+// See Directory.Entries for more
+type DirIterator interface {
+	// DirEntry holds information about current directory entry.
+	// Note that after creating new iterator you MUST call Next() at least once
+	// before accessing these methods. Calling these methods without prior calls
+	// to Next() and after Next() returned false may result in undefined behavior
+	DirEntry
 
-	Peek(n int) File
-	Length() int
+	// Next advances the iterator to the next file.
+	Next() bool
+
+	// Err may return an error after the previous call to Next() returned `false`.
+	// If the previous call to Next() returned `true`, Err() is guaranteed to
+	// return nil
+	Err() error
 }
 
-type SizeFile interface {
+// Directory is a special file which can link to any number of files.
+type Directory interface {
 	File
 
-	Size() (int64, error)
+	// Entries returns a stateful iterator over directory entries.
+	//
+	// Example usage:
+	//
+	// it := dir.Entries()
+	// for it.Next() {
+	//   name := it.Name()
+	//   file := it.File()
+	//   [...]
+	// }
+	// if it.Err() != nil {
+	//   return err
+	// }
+	//
+	// Note:
+	// - Below limitations aren't applicable to all implementations, consult
+	//   your implementations manual before using this interface in a way that
+	//   doesn't meet these constraints
+	// - Some implementations may only allow reading in order - so if the iterator
```

Comment:
We can probably loosen this restriction to:

1. Calling `Entries()` on a directory consumes it.
2. Calling `Next()` on a `DirIterator` consumes the previous entry.

That is, we can reasonably *allow* users to skip entire directories. Personally, with those restrictions, we can probably just make them a part of the API (they don't seem to onerous).

Reply:
Also, I take back what I said about unifying the MFS types and these, that's going to be *way* more work than it's worth and we'll probably hate ourselves for it.

I actually really like this streaming interface. I wish we had move semantics and borrows but I'm always going to miss rust a bit.

Reply:
So really this only applies to multipartfile as it's less than trivial to provide an esay way to skip stuff there, I guess I should probably move these notes there